File: app/src/main/java/edu/northeastern/group13project/PlayerHolder.java
package edu.northeastern.group13project;

import android.content.Context;
 

import androidx.annotation.NonNull;

 
import androidx.media3.exoplayer.ExoPlayer;

/**
 * Simple application-scoped holder for a single ExoPlayer instance so different Activities
 * can share playback state and UI.
 */
public class PlayerHolder {

    private static PlayerHolder instance;
    private final ExoPlayer player;

    private PlayerHolder(@NonNull Context appContext) {
        player = new ExoPlayer.Builder(appContext).build();
        // Do not set a default media item here; Activities will set MediaItem with proper metadata
        // -Note by Boxun :)
    }

    public static synchronized PlayerHolder getInstance(@NonNull Context context) {
        if (instance == null) {
            Context appContext = context.getApplicationContext();
            instance = new PlayerHolder(appContext);
        }
        return instance;
    }

    public ExoPlayer getPlayer() {
        return player;
    }

    public void release() {
        player.release();
    }
}

File: app/src/main/java/com/heartbeatmusic/PlayerHolder.kt
package com.heartbeatmusic

import android.content.Context
import androidx.media3.exoplayer.ExoPlayer

/**
 * Application-scoped holder for a single ExoPlayer instance so different
 * Activities can share playback state and UI.
 */
class PlayerHolder private constructor(appContext: Context) {

    val player: ExoPlayer = ExoPlayer.Builder(appContext).build()

    fun release() {
        player.release()
    }

    companion object {
        @Volatile
        private var instance: PlayerHolder? = null

        @Synchronized
        fun getInstance(context: Context): PlayerHolder {
            return instance ?: PlayerHolder(context.applicationContext).also { instance = it }
        }
    }
}
